src/core/string_map.cc

```diff
@@ -101,6 +101,36 @@ sds StringMap::Find(std::string_view key) {
   return GetValue(str);
 }
 
+sds StringMap::ReallocIfNeeded(void* obj, float ratio) {
+  sds key = (sds)obj;
+  size_t key_len = sdslen(key);
+
+  auto* value_ptr = key + key_len + 1;
+  uint64_t value_tag = absl::little_endian::Load64(value_ptr);
+  sds value = (sds)(uint64_t(value_tag) & kValMask);
+
+  // If the allocated value is underutilized, re-allocate it and update the pointer inside the key
+  if (ratio >= 1.0 || zmalloc_page_is_underutilized(value, ratio)) {
+    size_t value_len = sdslen(value);
+    sds new_value = sdsnewlen(value, value_len);
+    memcpy(new_value, value, value_len);
+    uint64_t new_value_tag = (uint64_t(new_value) & kValMask) | (value_tag & ~kValMask);
```

Comment:
`& kValMask` is redundant.

Reply:
Why? I want the value bits of one pointer and the tag bits if another.

Reply:
I see, because the upper unused bits are always sign extended... Still I don't see a problem with this slightly more explicit expression